### ISSUE TITLE
Directrtp broken

### ISSUE THREAD

**Issue by dkgroot** [marcelloceschia/chan-sccp-b#537]
Working on fixing directrtp

References: #405
References: #532
References: #394
References: #391

**Comment by drtmk** [comment]
After holding a call, I cannot resume it. Softkey doesn't do anything, and on the console I'm getting warnings:

```
[Sep 27 20:47:15] WARNING[8546][C-00000002]: sccp_session.c:281 sccp_session_waitForPendingRequests: SEPD0C282425243: waitForPendingRequests timed out!
    -- SEPD0C282425243: Set calledPartyNumber '120' calledPartyName 'Bob'
       > 0x7f93680378a0 -- Strict RTP learning after remote address set to: 10.100.0.202:29596
    -- SEPD0C282425243: Hold the channel SCCP/120-00000006
    -- SEPD0C282425243: Resume the channel SCCP/120-00000006
[Sep 27 20:47:20] WARNING[8546][C-00000002]: sccp_session.c:281 sccp_session_waitForPendingRequests: SEPD0C282425243: waitForPendingRequests timed out!
    -- SEPD0C282425243: Set calledPartyNumber '120' calledPartyName 'Bob'
    -- SEPD0C282425243: Hold the channel SCCP/120-00000006
    -- SEPD0C282425243: Resume the channel SCCP/120-00000006
[Sep 27 20:47:25] WARNING[8546][C-00000002]: sccp_session.c:281 sccp_session_waitForPendingRequests: SEPD0C282425243: waitForPendingRequests timed out!
    -- SEPD0C282425243: Set calledPartyNumber '120' calledPartyName 'Bob'
    -- SEPD0C282425243: Hold the channel SCCP/120-00000006
    -- SEPD0C282425243: Resume the channel SCCP/120-00000006
```

 ver. 4.3.2 issues/537 - 5fac962

**Comment by dkgroot** [comment]
@drtmk Hmmm, i did not get that, but i will check it out in a bit, to see what would cause this. I take it that a simple call using directrtp does now work again (correct ?).

**Comment by drtmk** [comment]
Before you made changes, I had no audio at all in both ways.
Now in a simple call scenario, audio works in both ways. I can hold a call and music on hold plays on the second phone, but I cannot resume it. Phone stops responding.

**Comment by dkgroot** [comment]
Can you set "sccp debug core, channel, rtp, pbx, socket" during testing ? I was not able to reproduce the 'resume' issue you saw, and you debug log was a little short for me to figure out what was going on in this instance.

**Comment by drtmk** [comment]
Right now when running  4.3.2 issues/537 - f415823, asterisk shuts down when phone is trying to register:

```
Asterisk Ready.
asterisk*CLI> sccp debug core, channel, rtp, pbx, socket
SCCP new debug status: (1 -> 12357) core,rtp,channel,pbx,socket
asterisk*CLI> sccp show version
Skinny Client Control Protocol (SCCP). Release: 4.3.2 issues/537 - f415823 (built by 'root' on '2020-09-30 06:15:44 UTC')
    -- SCCP: Accepted Client Connection from 10.100.0.200:50758
    -- SCCP: Connected on server via 10.100.0.2:2000
       > 10.100.0.2:2000: keepalive:62, keepaliveinterval:60
  == Alarm Type: LastOutOfServiceInformation
  == Reason Enum: 12
  == Last Event Sent: 16:CapabilitiesRes
  == Last Event Received: 155:CapabilitiesReq
    -- SCCP: (setOurIP4Address) client 10.100.0.200:50758
    -- SCCP: (setOurIP4Address) can be reached best via 10.100.0.2:2000
    -- SEP20370617CB3E: nat=Off set manually in config file
    -- SEP20370617CB3E: Request 'Capabilities Request' to device Pending
       > SEP20370617CB3E: keepalive:62, keepaliveinterval:59
    -- SEP20370617CB3E: Got rtpPort:3500 which the device wants to use for media
asterisk*CLI>
Disconnected from Asterisk server
Asterisk cleanly ending (0).
Executing last minute cleanups
Asterisk ending (0).
root@asterisk:/usr/src/directrtp#
```

I tried many times. Asterisk always shuts down.
I also restarted phones, to be sure they're not in some weird state.

Update:
When running asterisk -f, I can see an error:
```asterisk: sccp_utils.c:1504: sccp_copy_string: Assertion `NULL != dst && NULL != src' failed.```

**Comment by dkgroot** [comment]
Ooh crap i guess this might be related with work i did in the last couple of days. 
I just tested on mine, and it doesnt' crash here. 

Logically it must be one of these lines:
```
git diff d02d1d5 |grep sccp_copy
 			sccp_copy_string(buf, "--", buflen);
 				sccp_copy_string(buf, d->currentLine->id, buf_len);
 				sccp_copy_string(buf, d->pendingDelete ? "yes" : "no", buf_len);
-				sccp_copy_string(addrStr,sccp_netsock_stringify(&sas),sizeof(addrStr));								\
+				sccp_copy_string (addrStr, sccp_netsock_stringify (&sas), sizeof (addrStr)); \
-	sccp_copy_string(date, ctime(&t), sizeof(date));
-	sccp_copy_string(date, ctime(&t), sizeof(date));
+	sccp_copy_string (cntxt, GLOB (regcontext), sizeof (cntxt));
+			sccp_copy_string (cntxt, GLOB (regcontext), sizeof (cntxt));
 			sccp_copy_string(clientAddress, "--", sizeof(clientAddress));
```
Educated guess would point to "sccp_copy_string (cntxt, GLOB (regcontext), sizeof (cntxt));" as the culprit. It was using GLOB(regcontext) before checking if it was zero_length or not -> committed fix. 

If the problem persists on your side: Can you run with "sccp debug all" and see the last line calling sccp_copy_string before the crash ? or Can you show me the gdb backtrace for the core dump ?

**Comment by drtmk** [comment]
Asterisk works now.
So this is the log when Alice calls Bob. Bob holds a call, and tries to resume it.
[alice, bob, hold, resume.txt](https://github.com/chan-sccp/chan-sccp/files/5305578/alice.bob.hold.resume.txt)

I was able to resume it, but it worked after few seconds after pressing [Resume]. And audio worked only from Alice to Bob.
Same thing happens, when Alice holds a call and tries to resume it.

Basically when I hold a call, and I'm trying to resume it, phone stops responding. Like if it lost connection to asterisk.

**Comment by dkgroot** [comment]
@drtmk Thanks for checking and testing. I will have another look at the resume.txt to see what is different between your system and mine. The delay during resume is a bit of an issue, but i am not yet sure how to prevent it. During directrtp we need to start and new openReceiveChannel and have to wait for it to come back, to see the ip-address and port used, before we can continue the handling of resume.

**Comment by dkgroot** [comment]
I have changed the handling slightly. Have not looked at the one way audio for now, just trying to get rid of the delay first.

On my machine it looks like:
[directrtp_hold_resume.log](https://github.com/chan-sccp/chan-sccp/files/5306061/directrtp_hold_resume.log)
note1: i am performing hold/resume from the commandline, to show the actions taken
note2: my test server is pretty slow and the timing difference between our machines might be part of the problem you are seeing.

**Comment by dkgroot** [comment]
Pushed another update making call resume asynchronous, so the waitForPendingRequest can be processed in a separate thread. This should take care of the timeout events you were seeing. Apparently my machine was slow enough for the phones to beat the race condition :-)

**Comment by drtmk** [comment]
Good job. There's no delay right now. Phone behaves ok.
In a call transfer audio also works ok.
There's still a problem in one way audio after resume in cisco phones.
There's also one way audio between 79xx and SIP phone (using direct_media=yes in pjsip). Audio goes only from SIP to cisco. Direction of a call doesn't matter. Both phones are using g722.

Here's a log from a call from 100 (10.100.0.201 cisco) to 121 (SIP phone, 10.100.0.3).
[100 to 121 (SIP).txt](https://github.com/chan-sccp/chan-sccp/files/5309703/100.to.121.SIP.txt)

**Comment by dkgroot** [comment]
Ok will look into the one way issue. Looks like sip is declining the directrtp request while chan-sccp is still thinking that directrtp will succeed. The actual remote bridging is never started.

Also my solution to the delay, still needs a little tweaking

**Comment by dkgroot** [comment]
Made some smaller changes. Fixed the resume function so the channel lock was taken and released correctly. Updated codec framing and ip-address/port updating. Ran a couple of tests with chan_sip. which seemed ok.

When posting your debug log, please grab it from /var/log/asterisk/full with core set debug 3 set, so we can see how/when asterisk is performing the remote bridging part.

**Comment by dkgroot** [comment]
Still have not got resume back in working order. All the code is starting to get so convoluted :-(

**Comment by drtmk** [comment]
Don't put a pressure on yourself. Most people will not use directmedia. It's not a big deal.

**Comment by drtmk** [comment]
The only difference I see is that I can hear audio in both ways calling SIP phone. But I guess that doesn't use direct connection.
[sip.txt](https://github.com/chan-sccp/chan-sccp/files/5316251/sip.txt)
Any idea why they don't want to talk directly?

And here's a log from resume function taken from /var/log/asterisk/full. It's still one way audio.
[resume.txt](https://github.com/chan-sccp/chan-sccp/files/5316255/resume.txt)

**Comment by dkgroot** [comment]
I think i finally figured it out. and al lot of stuff i did in the last week might have been superfluous  but we will see. I will have a look in a couple of hours and see what can be done.

**Comment by dkgroot** [comment]
@drtmk 
Had to do quite a bit of moving code about and rethinking the solution to make it clean and nice again. Had to remove some of the earlier changes i made etc.
Have not dared to check how all these changes may have impacted "in-directrtp" to be honest :-). I hope i have done an ok job. I will check the fall out tomorrow.

**Comment by drtmk** [comment]
I will be in my phone lab on monday.

**Comment by dkgroot** [comment]
I had to test indirect-rtp with my new code, and low and behold it also works (with a tiny little patch)... => progress !

Now we just have to run a slew of test cases (simple call/hold/resume/park/transfer/pickup etc) to make sure they all work as well. Both direct and in-direct and see if we have any regressions.

actino | in-direct | direct | direct sip | direct + nat (remote office)
------- | ----- | ----- | ------- | -------
simple call | <ul><li>- [X] </li> | <ul><li>- [X] </li> | <ul><li>- [X] </li> | <ul><li>- [ ] </li>
hold / resume | <ul><li>- [X] </li> | <ul><li>- [X] </li> | <ul><li>- [X] </li> | <ul><li>- [ ] </li>
shared line hold / resume | <ul><li>- [ ] </li> | <ul><li>- [ ] </li> | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>
call forward | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>  | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>
transfer | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>  | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>
idivert / divert to voicemail | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>  | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>
monitor / record | <ul><li>- [ ] </li> | not supported | not supported | not supported
pickup | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>  | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>
gpickup | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>  | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>
park | <ul><li>- [X] </li> | <ul><li>- [X] </li> but missing park announcement | <ul><li>- [X] </li> | <ul><li>- [ ] </li>
barge | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>  | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>
conference | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>  | <ul><li>- [ ] </li> | <ul><li>- [ ] </li>

I am not sure we can easily support all this functionality with this current patch.

Note1: If one of these fails we need an excerpt from the "full" debug log file with:
- `sccp debug core, channel, rtp, pbx, rtp`
- `core set verbose 5`
- `core set debug 3`
set, so we can see the bridging work asterisk is doing.

**Comment by drtmk** [comment]
Using 4.3.2 issues/537 - 64a1b08 and directrtp enabled, there's no audio in both ways in a simple call.
But it starts working correctly when I make a hold and resume.
[call.txt](https://github.com/chan-sccp/chan-sccp/files/5325498/call.txt)

I would like to ask about above tests.
By call forward you mean use the CFwdALL softkey?
By conference you mean ConfBridge or that which is created from visual menu on the phone?
What kind of NAT whould you like to test? You want to place two phones behind NAT or Asterisk behind NAT?

I'm using asterisk 16.12.0

**Comment by dkgroot** [comment]
@drtmk I did not have simple call issues.

```
[Oct  5 07:52:49] VERBOSE[16873][C-00000008] :     -- SCCP: (pbx_indicate) start indicate 'AST_CONTROL_SRCCHANGE: Media source has changed and requires a new RTP SSRC'
[Oct  5 07:52:49] VERBOSE[16873][C-00000008] :     -- SCCP/100-00000008: (pbx_indicate) start indicate 'AST_CONTROL_SRCCHANGE: Media source has changed and requires a new RTP SSRC'
[Oct  5 07:52:49] VERBOSE[16873][C-00000008] :  SEPD0C28243AE44: (pbx_indicate) start indicate 'AST_CONTROL_SRCCHANGE: Media source has changed and requires a new RTP SSRC' (26) condition on channel SCCP/100-00000008 (rtp_instance:yes, reception.state:3/G.722 64k, transmission.state:0/G.722 64k)
[Oct  5 07:52:49] VERBOSE[16873][C-00000008] :   == SEPD0C28243AE44: (pbx_indicate) finish: send indication (res:0)
[Oct  5 07:52:49] VERBOSE[16874][C-00000008] :     -- SCCP: (pbx_indicate) start indicate 'AST_CONTROL_SRCCHANGE: Media source has changed and requires a new RTP SSRC'
[Oct  5 07:52:49] VERBOSE[16874][C-00000008] :     -- SCCP/120-00000009: (pbx_indicate) start indicate 'AST_CONTROL_SRCCHANGE: Media source has changed and requires a new RTP SSRC'
[Oct  5 07:52:49] VERBOSE[16874][C-00000008] :  SEPD0C282425243: (pbx_indicate) start indicate 'AST_CONTROL_SRCCHANGE: Media source has changed and requires a new RTP SSRC' (26) condition on channel SCCP/120-00000009 (rtp_instance:yes, reception.state:3/G.722 64k, transmission.state:0/G.722 64k)
```
Media transmission should have been started during the SRCCHANGE Phase (Which is the new way of doing things in the latest commits). Something seems to be preventing this.

Oooh i only made the changes for asterisk-16/17 and none of the others. Not sure what you are using ?

Pushed another change to alter the if condition in AST_CONTROL_SRCCHANGE handling.

BTW: from the call.txt file it's hard for me what type of bridging method is being used by asterisk (if it is attempting to do directrtp or not). That can only be seen in /var/log/asterisk/full log with `core set debug 3` set.

- yes cfwdall
- Conf softkey on the phone. Should (ought to) prevent directrtp and switch back to indirect automatically
- Single NAT on one side to start and then maybe two phones behind nat on a remote network, using directrtp between them (the classical case where directrtp makes the most sense).

**Comment by drtmk** [comment]
I will not have much time today to make these tests.

**Comment by dkgroot** [comment]
That last change i made is not a good idea. I have not slept yet :-(. I think your system is preventing directrtp during bringing, but we can only review that in the `full` log.
No problem. I need some sleep first anyway.

**Comment by dkgroot** [comment]
Figured out the cause, if the directrtp fails to be setup (for any cause) it will not fall back to in-direct like it should. Will need extra work. You can skip testing for now.
